File: python/tree/lowest-common-ancestor-of-a-binary-search-tree[1].py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        if min(p.val, q.val) > root.val:
            return self.lowestCommonAncestor(root.right, p, q)
        elif max(p.val, q.val) < root.val:
            return self.lowestCommonAncestor(root.left, p, q)
        else:
            return root

File: c/tree/lowest-common-ancestor-of-a-binary-search-tree[1].c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
#define min_2(x,y) ({ \
    int _x = (x); \
    int _y = (y); \
    _x < _y ? _x : _y;\
})

#define max_2(x,y) ({ \
    int _x = (x); \
    int _y = (y); \
    _x > _y ? _x : _y; \
})

struct TreeNode* lowestCommonAncestor(struct TreeNode* root, struct TreeNode* p, struct TreeNode* q) {
    if(min_2(p->val, q->val) > root->val)
        return lowestCommonAncestor(root->right, p, q);
    else if(max_2(p->val, q->val) < root->val)
        return lowestCommonAncestor(root->left, p, q);
    else
        return root;
}
